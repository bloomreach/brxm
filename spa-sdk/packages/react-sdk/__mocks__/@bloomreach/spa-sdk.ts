/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, Page } from '@bloomreach/spa-sdk';

const component = new class implements Component {
  getId = jest.fn();
  getMeta = jest.fn();
  getModels = jest.fn();
  getUrl = jest.fn();
  getName = jest.fn();
  getParameters = jest.fn();
  getChildren = jest.fn();
  getComponent = jest.fn();
  getComponentById = jest.fn();
};
const page = new class implements Page {
  getButton = jest.fn();
  getChannelParameters = jest.fn();
  getComponent = jest.fn(() => component);
  getContent = jest.fn();
  getDocument = jest.fn();
  getMeta = jest.fn();
  getTitle = jest.fn();
  getUrl = jest.fn();
  getVersion = jest.fn();
  getVisitor = jest.fn();
  getVisit = jest.fn();
  isPreview = jest.fn();
  rewriteLinks = jest.fn();
  sync = jest.fn();
  toJSON = jest.fn();
};

module.exports = {
  ...jest.genMockFromModule('@bloomreach/spa-sdk'),
  initialize: jest.fn(() => page),
};
