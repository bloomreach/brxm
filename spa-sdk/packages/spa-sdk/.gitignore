/coverage
/docs
/lib
/node_modules
