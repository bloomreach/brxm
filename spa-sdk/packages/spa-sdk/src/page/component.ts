/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { LinkFactory } from './link-factory';
import { Link } from './link';
import { MetaCollectionFactory } from './meta-collection-factory';
import { MetaCollectionModel, MetaCollection } from './meta-collection';
import { Reference } from './reference';

export const ComponentChildrenToken = Symbol.for('ComponentChildrenToken');
export const ComponentModelToken = Symbol.for('ComponentModelToken');

/**
 * Generic component type.
 */
export const TYPE_COMPONENT = 'component';

/**
 * Container item type.
 */
export const TYPE_COMPONENT_CONTAINER_ITEM = 'container-item';

/**
 * Container type.
 */
export const TYPE_COMPONENT_CONTAINER = 'container';

export type ComponentType = typeof TYPE_COMPONENT
  | typeof TYPE_COMPONENT_CONTAINER_ITEM
  | typeof TYPE_COMPONENT_CONTAINER;

/**
 * Meta-data of a component.
 */
export interface ComponentMeta extends MetaCollectionModel {
  params?: ComponentParameters;
}

type ComponentLinks = 'self';

type ComponentModels = Record<string, any>;

type ComponentParameters = Partial<Record<string, string>>;

/**
 * Model of a component.
 */
export interface ComponentModel {
  children?: Reference[];
  id: string;
  links: Record<ComponentLinks, Link>;
  meta: ComponentMeta;
  models?: ComponentModels;
  name?: string;
  type: ComponentType;
}

/**
 * A component in the current page.
 */
export interface Component {
  /**
   * @return The component id.
   */
  getId(): string;

  /**
   * @return The component meta-data collection.
   */
  getMeta(): MetaCollection;

  /**
   * @return The map of the component models.
   */
  getModels<T extends ComponentModels>(): T;

  /**
   * @return The link to the partial component model.
   */
  getUrl(): string | undefined;

  /**
   * @return The name of the component.
   */
  getName(): string;

  /**
   * @return The parameters of the component.
   */
  getParameters(): ComponentParameters;

  /**
   * @return The direct children of the component.
   */
  getChildren(): Component[];

  /**
   * Looks up for a nested component.
   * @param componentNames A lookup path.
   */
  getComponent<U extends Component>(...componentNames: string[]): U | undefined;
  getComponent(): this;

  /**
   * Looks up for a nested component by its id.
   * @param id A component id.
   */
  getComponentById<U extends Component>(id: string): U | this | undefined;
}

@injectable()
export class ComponentImpl implements Component {
  protected meta: MetaCollection;

  constructor(
    @inject(ComponentModelToken) protected model: ComponentModel,
    @inject(ComponentChildrenToken) protected children: Component[],
    @inject(LinkFactory) private linkFactory: LinkFactory,
    @inject(MetaCollectionFactory) metaFactory: MetaCollectionFactory,
  ) {
    this.meta = metaFactory(this.model.meta);
  }

  getId() {
    return this.model.id;
  }

  getMeta() {
    return this.meta;
  }

  getModels<T extends ComponentModels>(): T;
  getModels() {
    return this.model.models || {};
  }

  getUrl() {
    return this.linkFactory.create(this.model.links.self);
  }

  getName() {
    return this.model.name || '';
  }

  getParameters() {
    return this.model.meta.params || {};
  }

  getChildren() {
    return this.children;
  }

  getComponent(): this;
  getComponent<U extends Component>(...componentNames: string[]): U | undefined;
  getComponent(...componentNames: string[]) {
    // tslint:disable-next-line:no-this-assignment
    let component: Component | undefined = this;

    while (componentNames.length && component) {
      const name = componentNames.shift()!;
      component = component.getChildren().find(component => component.getName() === name);
    }

    return component;
  }

  getComponentById<U extends Component>(id: string): U | this | undefined;
  getComponentById(id: string) {
    const queue = [this as Component];

    while (queue.length) {
      const component = queue.shift()!;

      if (component.getId() === id) {
        return component;
      }

      queue.push(...component.getChildren());
    }
  }
}

/**
 * Checks whether a value is a page component.
 * @param value The value to check.
 */
export function isComponent(value: any): value is Component {
  return value instanceof ComponentImpl;
}
