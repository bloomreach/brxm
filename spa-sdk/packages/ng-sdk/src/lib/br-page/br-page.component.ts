/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  AfterContentChecked,
  AfterViewInit,
  ChangeDetectorRef,
  ChangeDetectionStrategy,
  Component,
  ContentChild,
  Inject,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  Output,
  Optional,
  PLATFORM_ID,
  SimpleChanges,
  TemplateRef,
  Type,
  ViewChild,
} from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { isPlatformBrowser, isPlatformServer } from '@angular/common';
import { makeStateKey, StateKey, TransferState } from '@angular/platform-browser';
import { from, of, BehaviorSubject, ReplaySubject, Subject } from 'rxjs';
import { filter, map, mapTo, pairwise, pluck, switchMap, take } from 'rxjs/operators';
import { destroy, initialize, isPage, Configuration, Page, PageModel } from '@bloomreach/spa-sdk';
import { BrComponentContext } from '../br-component.directive';
import { BrProps } from '../br-props.model';

interface BrNodeContext extends BrComponentContext {
  template?: TemplateRef<BrComponentContext>;
}

/**
 * The brXM page.
 */
@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'br-page',
  templateUrl: './br-page.component.html',
})
export class BrPageComponent implements AfterContentChecked, AfterViewInit, OnChanges, OnDestroy {
  /**
   * The configuration of the SPA SDK.
   * @see https://www.npmjs.com/package/@bloomreach/spa-sdk#configuration
   */
  @Input() configuration!: Omit<Configuration, 'httpClient'>;

  /**
   * The brXM and Angular components mapping.
   */
  @Input() mapping: Record<string, Type<BrProps>> = {};

  /**
   * The pre-initialized page instance or prefetched page model.
   * Mostly this property should be used to transfer state from the server-side to the client-side.
   */
  @Input() page?: Page | PageModel;

  /**
   * The TransferState key is used to transfer the state from the server-side to the client-side.
   * By default, it equals to `brPage`.
   * If `false` is passed then the state transferring feature will be disabled.
   */
  @Input() stateKey: StateKey<PageModel | undefined> | false = makeStateKey('brPage');

  /**
   * The current state of the page component.
   */
  @Output() state = new BehaviorSubject<Page | undefined>(undefined);

  @ViewChild('brNode') node!: TemplateRef<BrNodeContext>;

  @ContentChild(TemplateRef) private template?: TemplateRef<BrComponentContext>;

  private afterContentChecked$ = new Subject();

  private afterViewInit$ = new ReplaySubject();

  constructor(
    private changeDetectorRef: ChangeDetectorRef,
    private httpClient: HttpClient,
    @Inject(PLATFORM_ID) private platform: any,
    @Optional() private transferState?: TransferState,
  ) {
    this.request = this.request.bind(this);

    this.state.pipe(
      pairwise(),
      pluck(0),
      filter(isPage),
    )
    .subscribe(destroy);

    this.state.pipe(
      filter(isPage),
      switchMap((page) => this.afterContentChecked$.pipe(take(1), mapTo(page))),
    )
    .subscribe((page) => page.sync());

    this.state.pipe(
      filter(() => isPlatformServer(this.platform)),
      filter(isPage),
    )
    .subscribe((page) => this.stateKey && this.transferState?.set(this.stateKey, page.toJSON()));
  }

  get context(): BrNodeContext | undefined {
    const page = this.state.getValue();
    const component = page?.getComponent();

    if (!page || !component) {
      return;
    }

    return {
      component,
      page,
      $implicit: component,
      template: this.template,
    };
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes.configuration || changes.page) {
      this.initialize(changes.page?.currentValue);
    }

    if (changes.stateKey?.previousValue && isPlatformServer(this.platform)) {
      if (changes.stateKey.currentValue && this.transferState?.hasKey(changes.stateKey.previousValue)) {
        this.transferState?.set(
          changes.stateKey.currentValue,
          this.transferState?.get(changes.stateKey.previousValue, undefined),
        );
      }

      this.transferState?.remove(changes.stateKey.previousValue);
    }
  }

  ngAfterContentChecked(): void {
    this.afterContentChecked$.next();
  }

  ngAfterViewInit(): void {
    this.afterViewInit$.next();
  }

  ngOnDestroy(): void {
    this.state.next(undefined);
    this.state.complete();
    this.afterContentChecked$.complete();
    this.afterViewInit$.complete();
  }

  private initialize(page: Page | PageModel | undefined): void {
    if (this.stateKey && isPlatformBrowser(this.platform) && this.transferState?.hasKey(this.stateKey)) {
      page = page ?? this.transferState?.get(this.stateKey, undefined);
      this.transferState?.remove(this.stateKey);
    }

    const configuration = { httpClient: this.request, ...this.configuration } as Configuration;
    const observable = page
      ? of(initialize(configuration, page))
      : from(initialize(configuration));

    observable.pipe(
      switchMap(state => this.afterViewInit$.pipe(take(1), mapTo(state))),
    )
    .subscribe(state => {
      this.state.next(state);
      this.changeDetectorRef.detectChanges();
    });
  }

  private request(...[{ data: body, headers, method, url }]: Parameters<Configuration['httpClient']>) {
    return this.httpClient.request<PageModel>(
      method,
      url,
      {
        body,
        headers: headers as Record<string, string | string[]>,
        responseType: 'json'
      },
    )
    .pipe(map(data => ({data})))
    .toPromise();
  }
}
