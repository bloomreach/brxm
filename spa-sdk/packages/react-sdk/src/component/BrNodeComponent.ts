/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Component } from '@bloomreach/spa-sdk';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';

export class BrNodeComponent<T extends Component> extends React.Component<BrProps<T>> {
  static contextType = BrMappingContext;
  context!: React.ContextType<typeof BrMappingContext>;

  protected getMapping(): string | undefined {
    return this.props.component.getName();
  }

  protected fallback() {
    return this.props.children;
  }

  render() {
    const mapping = this.getMapping();
    const component = mapping && this.context[mapping];
    if (!component) {
      return this.fallback();
    }

    return React.createElement(component, this.props);
  }
}
