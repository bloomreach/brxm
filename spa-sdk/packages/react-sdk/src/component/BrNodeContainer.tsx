/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  Container,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrNodeComponent } from './BrNodeComponent';
import { BrContainerBox, BrContainerInline, BrContainerNoMarkup, BrContainerOrderedList, BrContainerUnorderedList } from '../cms';

export class BrNodeContainer extends BrNodeComponent<Container> {
  protected getMapping() {
    return this.props.component.getType();
  }

  protected fallback() {
    switch (this.props.component.getType()) {
      case TYPE_CONTAINER_INLINE: return <BrContainerInline {...this.props} />;
      case TYPE_CONTAINER_NO_MARKUP: return <BrContainerNoMarkup {...this.props} />;
      case TYPE_CONTAINER_ORDERED_LIST: return <BrContainerOrderedList {...this.props} />;
      case TYPE_CONTAINER_UNORDERED_LIST: return <BrContainerUnorderedList {...this.props} />;
      default: return <BrContainerBox {...this.props} />;
    }
  }
}
