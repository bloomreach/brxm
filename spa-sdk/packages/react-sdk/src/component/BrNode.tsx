/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isContainer, isContainerItem, Component } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';
import { BrComponentContext } from './BrComponentContext';
import { BrNodeContainer } from './BrNodeContainer';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { BrNodeComponent } from './BrNodeComponent';

interface BrNodeProps {
  component: Component;
}

export class BrNode extends React.Component<BrNodeProps> {
  static contextType = BrPageContext;
  context: React.ContextType<typeof BrPageContext>;

  private renderNode() {
    if (React.Children.count(this.props.children)) {
      return this.props.children;
    }

    const children = this.props.component.getChildren()
      .map((child, index) => <BrNode key={index} component={child} />);

    if (isContainer(this.props.component)) {
      return (
        <BrNodeContainer component={this.props.component} page={this.context!}>
          {children}
        </BrNodeContainer>
      );
    }

    if (isContainerItem(this.props.component)) {
      return (
        <BrNodeContainerItem component={this.props.component} page={this.context!}>
          {children}
        </BrNodeContainerItem>
      );
    }

    return (
      <BrNodeComponent component={this.props.component} page={this.context!}>
        {children}
      </BrNodeComponent>
    );
  }

  render() {
    return(
      <BrComponentContext.Provider value={this.props.component}>
        <BrMeta meta={this.props.component.getMeta()}>
          {this.renderNode()}
        </BrMeta>
      </BrComponentContext.Provider>
    );
  }
}
