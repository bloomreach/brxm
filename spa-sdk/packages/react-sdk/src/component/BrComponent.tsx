/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { BrComponentContext } from './BrComponentContext';
import { BrNode } from './BrNode';

interface BrComponentProps {
  /**
   * The path to a component.
   * The path is defined as a slash-separated components name chain
   * relative to the current component (e.g. `main/container`).
   * If it is omitted, all the children will be rendered.
   */
  path?: string;
}

/**
 * The brXM component.
 */
export class BrComponent extends React.Component<BrComponentProps> {
  static contextType = BrComponentContext;
  context: React.ContextType<typeof BrComponentContext>;

  private getComponents() {
    if (!this.context) {
      return [];
    }
    if (!this.props.path) {
      return this.context.getChildren();
    }

    const component = this.context.getComponent(...this.props.path.split('/'));

    return component ? [component] : [];
  }

  private renderComponents() {
    return this.getComponents().map((component, index) => (
      <BrNode key={index} component={component}>{this.props.children}</BrNode>
    ));
  }

  render() {
    return <>{this.renderComponents()}</>;
  }
}
