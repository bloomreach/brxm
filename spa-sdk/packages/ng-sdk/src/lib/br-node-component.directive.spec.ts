/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, ElementRef, Input, NgModule, TemplateRef, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { Component as SpaComponent, MetaCollection, Page } from '@bloomreach/spa-sdk';
import { BrComponentContext } from './br-component.directive';
import { BrNodeComponentDirective } from './br-node-component.directive';
import { BrNodeDirective } from './br-node.directive';
import { BrPageComponent } from './br-page/br-page.component';
import { BehaviorSubject } from 'rxjs';

@Component({
  selector: 'br-component-test',
  template: `
    <span>Component: {{ component.getName() }}</span>
    <span>Page: {{ page.getTitle() }}</span>
  `,
})
class ComponentTestComponent {
  @Input() component!: SpaComponent;
  @Input() page!: Page;
}

@NgModule({
  declarations: [ComponentTestComponent],
  entryComponents: [ComponentTestComponent],
})
class TestModule {}

@Component({
  template: `
    <ng-template #template let-implicit let-component="component" let-page="page">
      <span class="head">Implicit: {{ implicit.getName() }}</span>
      <span>Component: {{ component.getName() }}</span>
      <span class="tail">Page: {{ page.getTitle() }}</span>
    </ng-template>
  `,
})
class TemplateComponent {
  @ViewChild('template') template!: TemplateRef<BrComponentContext>;
}

@Component({ template: '<ng-container #container [brNodeComponent]="component"></ng-container>' })
class TestComponent {
  @Input() component!: SpaComponent;
  @ViewChild('container') container!: ElementRef<Comment>;
}

describe('BrNodeComponentDirective', () => {
  let clear: jest.Mocked<ReturnType<MetaCollection['render']>>;
  let component: jest.Mocked<SpaComponent>;
  let meta: jest.Mocked<MetaCollection>;
  let node: BrNodeDirective;
  let page: BrPageComponent;
  let template: TemplateRef<BrComponentContext>;
  let fixture: ComponentFixture<TestComponent>;

  beforeEach(() => {
    clear = jest.fn();
    meta = { render: jest.fn(() => clear) } as unknown as typeof meta;
    component = {
      getChildren: jest.fn(() => []),
      getMeta: jest.fn(() => meta),
      getName: jest.fn(() => 'something'),
    } as unknown as typeof component;
    node = {} as BrNodeDirective;
    page = {
      mapping: {},
      state: new BehaviorSubject({ getTitle: () => 'Some Page' }),
    } as unknown as typeof page;
  });

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ BrNodeComponentDirective, TemplateComponent, TestComponent ],
      imports: [ TestModule ],
      providers: [
        { provide: BrNodeDirective, useFactory: () => node },
        { provide: BrPageComponent, useFactory: () => page },
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    const templateFixture = TestBed.createComponent(TemplateComponent);
    templateFixture.detectChanges();
    ({ template } = templateFixture.componentInstance);

    fixture = TestBed.createComponent(TestComponent);
    fixture.componentInstance.component = component;
  });

  describe('ngOnChanges', () => {
    it('should render a node template', () => {
      node.template = template;
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
      expect(meta.render).toBeCalledWith(
        fixture.nativeElement.querySelector('.head'),
        fixture.nativeElement.querySelector('.tail'),
      );
    });

    it('should render children when there is no mapping', () => {
      const child1 = { getName: () => 'child1' } as SpaComponent;
      const child2 = { getName: () => 'child2' } as SpaComponent;
      component.getChildren.mockReturnValue([child1, child2]);
      page.node = template;
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
      expect(meta.render).toBeCalledWith(
        fixture.nativeElement.querySelectorAll('.head')[0],
        fixture.nativeElement.querySelectorAll('.tail')[1],
      );
    });

    it('should surround the container with meta-data when there is no children to render', () => {
      fixture.detectChanges();

      expect(meta.render).toBeCalledWith(
        fixture.componentInstance.container.nativeElement,
        fixture.componentInstance.container.nativeElement,
      );
    });

    it('should render a mapped component', () => {
      page.mapping.something = ComponentTestComponent;
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
      expect(meta.render).toBeCalledWith(
        fixture.nativeElement.querySelector('br-component-test'),
        fixture.nativeElement.querySelector('br-component-test'),
      );
    });
  });

  describe('ngOnDestroy', () => {
    it('should clear the container', () => {
      node.template = template;
      fixture.detectChanges();
      fixture.destroy();

      expect(fixture.nativeElement).toMatchSnapshot();
      expect(clear).toBeCalled();
    });
  });
});
