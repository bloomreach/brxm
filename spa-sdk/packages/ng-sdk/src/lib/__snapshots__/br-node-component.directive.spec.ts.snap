// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNodeComponentDirective ngOnChanges should render a mapped component 1`] = `
<div
  id="root7"
>
  
  <br-component-test>
    <span>
      Component: something
    </span>
    <span>
      Page: Some Page
    </span>
  </br-component-test>
</div>
`;

exports[`BrNodeComponentDirective ngOnChanges should render a node template 1`] = `
<div
  id="root1"
>
  
  <span
    class="head"
  >
    Implicit: something
  </span>
  <span>
    Component: something
  </span>
  <span
    class="tail"
  >
    Page: Some Page
  </span>
</div>
`;

exports[`BrNodeComponentDirective ngOnChanges should render children when there is no mapping 1`] = `
<div
  id="root3"
>
  
  <span
    class="head"
  >
    Implicit: child1
  </span>
  <span>
    Component: child1
  </span>
  <span
    class="tail"
  >
    Page: Some Page
  </span>
  <span
    class="head"
  >
    Implicit: child2
  </span>
  <span>
    Component: child2
  </span>
  <span
    class="tail"
  >
    Page: Some Page
  </span>
</div>
`;

exports[`BrNodeComponentDirective ngOnDestroy should clear the container 1`] = `
<div
  id="root9"
>
  
</div>
`;
